File: dashboards/ruby/routes/vault.rb
# frozen_string_literal: true

# Vault routes — secrets management, rotate, delete

get '/vault' do
  require_auth!
  res = api_get('/api/engine/vault/secrets?orgId=default')
  @secrets = res[:status] == 200 ? (res[:body]['secrets'] || res[:body]) : []
  @secrets = @secrets.is_a?(Array) ? @secrets : []
  erb :vault
end

post '/vault' do
  require_auth!
  res = api_post('/api/engine/vault/secrets', {
    name: params[:name],
    value: params[:value],
    category: params[:category] || 'custom',
    orgId: 'default'
  })
  if res[:status] < 300
    set_flash('Secret added successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to add secret') : 'Failed to add secret', 'danger')
  end
  redirect '/vault'
end

post '/vault/:id/delete' do
  require_auth!
  res = api_delete("/api/engine/vault/secrets/#{params[:id]}")
  if res[:status] < 300
    set_flash('Secret deleted successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to delete secret') : 'Failed to delete secret', 'danger')
  end
  redirect '/vault'
end

post '/vault/:id/rotate' do
  require_auth!
  res = api_post("/api/engine/vault/secrets/#{params[:id]}/rotate", {})
  if res[:status] < 300
    set_flash('Secret encryption rotated successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to rotate secret') : 'Failed to rotate secret', 'danger')
  end
  redirect '/vault'
end

File: dashboards/python/routes/vault.py
"""
Vault routes: manage secrets — add, rotate, and delete.
"""

from flask import Blueprint, render_template, request, session, redirect, url_for, flash
from functools import wraps
from utils.api import api_request

vault_bp = Blueprint('vault_bp', __name__)


def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'token' not in session:
            return redirect(url_for('auth_bp.login'))
        return f(*args, **kwargs)
    return decorated


@vault_bp.route('/vault')
@login_required
def vault():
    data = api_request('GET', '/engine/vault/secrets?orgId=default')
    return render_template(
        'vault.html',
        active_page='vault',
        user=session.get('user', {}),
        secrets=data.get('secrets', []),
    )


@vault_bp.route('/vault/add', methods=['POST'])
@login_required
def add_secret():
    body = {
        'name': request.form['name'],
        'value': request.form['value'],
        'category': request.form.get('category', 'custom'),
    }
    result = api_request('POST', '/engine/vault/secrets', body)
    flash('Secret added!' if 'id' in result else result.get('error', 'Failed'))
    return redirect(url_for('vault_bp.vault'))


@vault_bp.route('/vault/<id>/rotate', methods=['POST'])
@login_required
def rotate_secret(id):
    result = api_request('POST', f'/engine/vault/secrets/{id}/rotate')
    flash('Secret rotated!' if not result.get('error') else result.get('error', 'Failed'))
    return redirect(url_for('vault_bp.vault'))


@vault_bp.route('/vault/<id>/delete', methods=['POST'])
@login_required
def delete_secret(id):
    api_request('DELETE', f'/engine/vault/secrets/{id}')
    flash('Secret deleted')
    return redirect(url_for('vault_bp.vault'))
